grant-project/zcash-grant-system
# Admin users

Closes #210. Closes #193.

Replaces static env based single admin, with user based admin roles. Includes TOTP 2FA with backup codes.

### What 
- remove old single admin auth
- adds `is_admin` to user table, and adds toggle to admin user-detail view to set it
- adds `backup_codes` to user table, holds flask-sec hashed comma separated recovery codes
- adds 2FA setup for admin users at login time if it is not set up
- adds "settings" page to admin, just has a button to reset 2FA
- allows admin users to reset 2FA
- adds new python dep `pyotp==2.2.7`
- adds admin deps `qrcode.react` and `react-copy-to-clipboard`
- adds decent test of the flow

### Notes
- recovery codes are checked if the attempt to verify the totp code fails, it does so by going through the list of hashes, this checking is somewhat slow -- only an admin user who has already set up 2fa can attempt to use recovery codes, so I do not suspect adversarial behavior here, open to thoughts

### What's not here
- no emails included here, I suspect we'll want to get an admin email template and send emails for certain situations (not just auth related)


### Process (new admin)
1. User visits admin site, if not authenticated on frontend or admin, will be asked for login
1. First login (no 2FA setup)
    1. User will be required to go throw the setup process before proceeding
    1. If user last login event > 20 minutes old, user will be asked for password
    1. User will receive recovery codes and totp secret from server
    1. User asked to save recovery codes
    1. User sets up Authenticator app, and enters current code
    1. User sends back backup codes and secret
        1. Server checks hash of settings (in lieu of storing unverified settings in the DB)
        1. Server validates the user generated validation code using the secret
        1. Server stores secret and backup codes to user table
    1. User should now have full access to admin site
1. Subsequent logins
    1. If user is already authenticated on frontend
        1. User will be asked for totp code
    1. If user not authenticated on frontend
        1. User enter username and password
        1. User enters totp code
    1. User now authenticated on admin


### Tests

Sorry for the jumble, but here goes.

`admin/> yarn`
`backend/> flask db upgrade` 🤞 

NOTE: Set first admin manually by setting `user.is_admin=TRUE` in DB.

- user, not admin, not verified
    1. attempt to login using email & password
    1. should get "user or pass incorrect"

- user, admin, not verified
    1. login on frontend or admin using email/pass 
    1. should get error alert explaining things

- user, admin, verified email, stale login
    ```py
    # grant/utils/auth.py:31
    def is_auth_fresh(minutes: int=1):
    ```
    1. set minutes to 1
    1. if not logged in, do so in admin
    1. should get 2FA setup screen
    1. wait > 1 minute, reload page, should get "Please verify your password"
    1. (rest of 2fa initial setup)

- user, admin, verified email, fresh login
    1. login to admin
    1. (rest of 2fa initial setup)

- settings > Setup 2FA, stale
    1. go through all the setup steps and be logged in, make sure the fresh timeout is set to 1 minute
    1. allow a minute to pass
    1. visit Settings > Setup 2FA, you should be asked for password before proceeding to 2FA setup

- general 2FA setup tests
    1. cancel button should take user back to first message on initial setup
    1. cancel button should take user back to Settings page when creating new 2FA setup
    1. should not be able to initiate setup process when login is "stale"
    1. should be able to download a text file of recovery codes
    1. should be able to copy-paste recovery codes
    1. should be able to scan code (I've testing with Google Authenticator)
    1. should be able to click "enter the text code" or the T icon on the card footer to see secret totp code
    1. totp code should allow generator to generate correct codes

- recovery codes
    1. should be able to use recovery codes one time only
    1. should be able to omit '-' from recovery code
    1. when remaining codes count < 5, should get a warning when verifying code on login suggesting setting new codes via setup process

- should be able to toggle user admin state via actions sidebar in user details

## Screens

### Login
![screen shot 2019-02-22 at 9 26 52 am](https://user-images.githubusercontent.com/11430954/53260367-ee266a00-3696-11e9-9ca5-bc95af0974f6.png)

### Login - is admin - email not verified
![screen shot 2019-02-22 at 10 48 20 am](https://user-images.githubusercontent.com/11430954/53260449-229a2600-3697-11e9-9aef-17dfdea54165.png)


### Login stale - password verify
![screen shot 2019-02-22 at 10 55 07 am](https://user-images.githubusercontent.com/11430954/53260395-01d1d080-3697-11e9-8d74-fde137a910f5.png)

### Initial setup - 1. Outline
![screen shot 2019-02-22 at 9 27 24 am](https://user-images.githubusercontent.com/11430954/53260484-38a7e680-3697-11e9-96a8-48f012b339d1.png)

### Initial setup - 2. Save Recovery Codes
![screen shot 2019-02-22 at 9 27 43 am](https://user-images.githubusercontent.com/11430954/53260514-49585c80-3697-11e9-89a2-5dde4aa4b7af.png)

### Initial setup - 2. Scan Code & Verify
![screen shot 2019-02-22 at 9 34 32 am](https://user-images.githubusercontent.com/11430954/53260543-58d7a580-3697-11e9-9864-2c60e9c19b09.png)

### Login, 2nd Factor
![screen shot 2019-02-22 at 9 36 07 am](https://user-images.githubusercontent.com/11430954/53260571-7147c000-3697-11e9-825f-b53e548005d3.png)

### Login, 2nd Factor - low recovery codes
![screen shot 2019-02-22 at 9 43 43 am](https://user-images.githubusercontent.com/11430954/53260614-8a507100-3697-11e9-948d-403ad314c8a9.png)

### Admin > Settings
![screen shot 2019-02-22 at 10 29 11 am](https://user-images.githubusercontent.com/11430954/53260633-950b0600-3697-11e9-9491-9b72af2ee3f6.png)
